DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,412 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace DataStructures.BinarySearchTree
+{
+    /// <summary>
+    /// An ordered tree with efficient insertion, removal, and lookup.
+    /// </summary>
+    /// <remarks>
+    /// A Binary Search Tree (BST) is a tree that satisfies the following properties:
+    /// <list type="bullet">
+    /// <item>All nodes in the tree contain two children, usually called Left and Right.</item>
+    /// <item>All nodes in the Left subtree contain keys that are less than the node's key.</item>
+    /// <item>All nodes in the Right subtree contain keys that are greater than the node's key.</item>
+    /// </list>
+    /// A BST will have an average complexity of O(log n) for insertion, removal, and lookup operations.
+    /// </remarks>
+    /// <typeparam name="TKey">Type of key for the BST. Keys must implement IComparable.</typeparam>
+    public class BinarySearchTree<TKey> where TKey : IComparable<TKey>
+    {
+        /// <summary>
+        /// Private class member to hold the root of the BST.
+        /// </summary>
+        private BinarySearchTreeNode<TKey>? root;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinarySearchTree{TKey}"/> class.
+        /// </summary>
+        public BinarySearchTree()
+        {
+            root = null;
+            Count = 0;
+        }
+
+        /// <summary>
+        /// Gets the number nodes currently in the BST.
+        /// </summary>
+        public int Count { get; private set; }
+
+        /// <summary>
+        /// Insert a key into the BST.
+        /// </summary>
+        /// <param name="key">The key to insert.</param>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Thrown if key is already in BST.
+        /// </exception>
+        public void Add(TKey key)
+        {
+            if (Equals(key, default))
```

Comment:
Can we add 0 when TKey is int? Would `is null` check work?
```suggestion
            if (key is null)
```